### PR TITLE
Client Resiliency

### PR DESCRIPTION
**Pull request by abhishek58g**
Retry logic taking into account client metrics.

### CODE REVIEW

File: src/ADAL.PCL/AdalHttpClient.cs [hunk 1]
```diff
@@ -122,7 +122,69 @@ public AdalHttpClient(string uri, CallState callState)
...
+                    {
+                        response = ex.WebResponse;
+                    }
+                }
+                finally
+                {
+                    clientMetricsRetry.EndClientMetricsRecord(endpointType, this.CallState);
+                }
+            }
+
```
Comment (kpanwar) [inline src/ADAL.PCL/AdalHttpClient.cs]: this is all copy paste code. Consider recursive call.

File: src/ADAL.PCL/AuthenticationResult.cs [hunk 2]
```diff
@@ -74,6 +74,19 @@ internal AuthenticationResult(string accessTokenType, string accessToken, DateTi
         public DateTimeOffset ExpiresOn { get; internal set; }
 
         /// <summary>
+        /// Gets the point in time in which the Access Token returned in the AccessToken property ceases to be valid in ADAL's extended LifeTime.
+        /// This value is calculated based on current UTC time measured locally and the value ext_expiresIn received from the service.
+        /// </summary>
+        [DataMember]
+        public DateTimeOffset ExtendedExpiresOn { get; internal set; }
```
Comment (kpanwar) [inline src/ADAL.PCL/AuthenticationResult.cs]: why is this public? why should a developer care when there is a Boolean already that denotes it?

File: src/ADAL.PCL/HttpClientWrapper.cs [hunk 3]
```diff
@@ -119,6 +122,23 @@ public int TimeoutInMilliSeconds
 
                 IHttpWebResponse webResponse = await CreateResponseAsync(responseMessage);
 
+                // Retry the request here
+                if ((int) webResponse.StatusCode == 500 || (int) webResponse.StatusCode == 503 ||
+                    (int) webResponse.StatusCode == 504)
+                {
+                    await Task.Delay(TimeSpan.FromSeconds(1));
```
Comment (kpanwar) [inline src/ADAL.PCL/HttpClientWrapper.cs]: again, copy paste code.

File: src/ADAL.PCL/HttpClientWrapper.cs [hunk 4]
```diff
@@ -119,6 +122,23 @@ public int TimeoutInMilliSeconds
 
                 IHttpWebResponse webResponse = await CreateResponseAsync(responseMessage);
 
+                // Retry the request here
+                if ((int) webResponse.StatusCode == 500 || (int) webResponse.StatusCode == 503 ||
```
Comment (kpanwar) [inline src/ADAL.PCL/HttpClientWrapper.cs]: why is error check in 2 places?

File: src/ADAL.PCL/AcquireTokenHandlerBase.cs [hunk 5]
```diff
@@ -105,9 +105,15 @@ internal abstract class AcquireTokenHandlerBase
...
         protected bool LoadFromCache { get; set; }
-        
+
         protected bool StoreToCache { get; set; }
 
+        private const int ExpirationMarginInMinutes = 5;
+
+        private bool _tokenNearExpiry;
+
+        private bool _tokenExtendedLifeTimeExpired;
```
Comment (kpanwar) [inline src/ADAL.PCL/AcquireTokenHandlerBase.cs]: why does acquire token handler care about extended life time? all the logic should be hidden inside token cache.

File: src/ADAL.PCL/AcquireTokenHandlerBase.cs [hunk 6]
```diff
@@ -105,9 +105,15 @@ internal abstract class AcquireTokenHandlerBase
         protected UserIdentifierType UserIdentifierType { get; set; }
 
         protected bool LoadFromCache { get; set; }
-        
+
         protected bool StoreToCache { get; set; }
 
+        private const int ExpirationMarginInMinutes = 5;
```
Comment (kpanwar) [inline src/ADAL.PCL/AcquireTokenHandlerBase.cs]: duplicate variable. handlers should not care about the margin

File: src/ADAL.PCL/AcquireTokenHandlerBase.cs [hunk 7]
```diff
@@ -105,9 +105,15 @@ internal abstract class AcquireTokenHandlerBase
...
+
+        private bool _tokenNearExpiry;
```
Comment (kpanwar) [inline src/ADAL.PCL/AcquireTokenHandlerBase.cs]: handlers should not care id the token is near expiry. cache either returns it, or not.

File: src/ADAL.PCL/AdalHttpClient.cs (hunk 1 above)
Comment (abhishek58g) [inline src/ADAL.PCL/AdalHttpClient.cs]: Made a recursive call

**Pull request closed without merging**